### ISSUE TITLE
Eliminate the need for inheritance for action controllers

### ISSUE THREAD

**Issue by magento-engcom-team** [magento-engcom/community-features#9]
By design, all Magento action controllers should implement `\Magento\Framework\App\ActionInterface`. But some crucial request processing behavior (event dispatching, authorisation, etc) resides in classes like `\Magento\Framework\App\Action\Action` and `\Magento\Backend\App\AbstractAction`. So if an action controller implements the `ActionInterface` but does not inherit from one of these "layer supertype" classes, it will lose that crucial behavior.

So, Magento module developers do not have a way to create an action controller without using inheritance. To avoid inheritance ([see why](http://antonkril.github.io/code-reuse)), the request processing behavior should be extracted from "layer supertypes".  Proposed solution is to move the behavior to action controller plugins.

Action controller should contain only its custom behavior.

AC:
- module developer does not have to extend from any class to create a fully functional action controller, implementing `\Magento\Framework\App\ActionInterface` is enough
- controllers "supertypes" (`\Magento\Backend\App\AbstractAction`, `\Magento\Framework\App\Action\Action`, `\Magento\Framework\App\Action\AbstractAction`) are deprecated
- magento supports both controller implementations (inheritance based and non-inheritance based) 
- [optional] controllers are migrated

**Original Report**: https://github.com/magento/magento2/issues/9582 by @antonkril

**Comment by ihor-sviziev** [comment]
Basically this feature request was already implemented
- ✔ implemented in https://github.com/magento/magento2/pull/26778 (thanks @vinai and @lbajsarowicz),
- ✔ documentation updated in https://github.com/magento/devdocs/pull/6901 (thanks @lbajsarowicz)
- ✔ new controllers should not use inheritance - enforced by static test in https://github.com/magento/magento2/pull/29339 (thanks @swnsma).

The only thing that still not done 
- ❌controllers still weren't migrated to not use inheritance

Good job! 👍

**Comment by Vinai** [comment]
Migrating controllers would be an unnecessary backward compatibility break.
Why not keep them the way they are, and only enforce new ones to use composition?

**Comment by ihor-sviziev** [comment]
@Vinai I would agree with you that migration of all controllers might be not necessary, but I'm strongly believe we should have good examples as much as possible in order to show to extension developers how to use controllers without inheritance and that it's best practice.

Also for some controllers it might be really needed, for instance for controller that handles ESI requests from varnish.

**Comment by mannebusk** [comment]
Was using "controller_action_predispatch" to forward requests, by doing something like this
```
$request->initForward()
            ->setRouteName('routename')
            ->setModuleName('modulename')
            ->setControllerName('index')
            ->setActionName('index')
            ->setDispatched(false);
```
This does not work any more since the event dispatching was moved to a plugin.

Before, the original matched actions `execute()` was never run, because it checked if request was dispatched before running it. Now `execute()` is always run.

Just wanted to report, in case it wasn't intentional to break it. And if some one else have the same problem, solved it by making a Router to handle forwarding instead.

**Comment by mrtuvn** [comment]
Have you test your case again in latest code branch (2.4-develop github). New approach no-longer used plugin for event dispatch
`controller_action_predispatch`

**Comment by Sam-Mear** [comment]
Was backend/admin controllers ever updated to eliminate inheritance?

**Comment by ihor-sviziev** [comment]
@Sam-Mear, unfortunately, no :(

**Comment by lbajsarowicz** [comment]
@Sam-Mear No, these were not. The current structure of Admin controllers was too dependent of abstract structure to do it. And now it does not even make sense, because... there's no team to review and approve it on Adobe side.

**Comment by Sam-Mear** [comment]
> @Sam-Mear No, these were not. The current structure of Admin controllers was too dependent of abstract structure to do it. And now it does not even make sense, because... there's no team to review and approve it on Adobe side.


Theres no team to review and approve it on Adobe side? Do Adobe touch the open source side? Looks like all the pull requests either get ignored or maintained by Magento partners

**Comment by sprankhub** [comment]
With the community prioritisation process, we can pick the PRs that are being processed. So it would be possible to get this checked.
Additionally, the Magento Association is actively working on improving the PR process again. Improvements are coming :)

**Comment by lbajsarowicz** [comment]
@sprankhub In the past, there was a group of people who thought outside of procedures in order to implement project-wide changes - e.g. PHPUnit upgrade from 6 to 9, giving me a dedicated PoC inside of the organization and access to the Commerce packages to modify these and create a bulk of PRs that are merged all together. 

Right now I don't see any opportunity to do the same for sake of project-wide changes, especially given that Adobe is moving towards App Builder.

**Comment by Den4ik** [comment]
Many years have passed, but there still isn’t a single example in the core of moving away from inheritance.
Is this idea still alive, or has it been die?

**Comment by Vinai** [comment]
The existing controllers should stay the way they are - only new controllers can use the interface based approach.
Refactoring existing controllers would be a huge BC break, without any real benefits.

**Comment by Sam-Mear** [comment]
> The existing controllers should stay the way they are - only new controllers can use the interface based approach.
> Refactoring existing controllers would be a huge BC break, without any real benefits.

The benefit is anyway up to date codebase. Breaking changes are fine if versioned

**Comment by Vinai** [comment]
> Breaking changes are fine if versioned

Strongly disagree. I'l won't argue further on this thread though.

**Comment by Den4ik** [comment]
@Vinai, I agree with you about the old controllers. However, it looks like all new controllers added over the 8 years since this topic was first discussed still follow the old style. None of them were added without using inheritance.
